src/net.cpp

```diff
@@ -1152,6 +1152,37 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     RandAddEvent((uint32_t)id);
 }
 
+bool CConnman::AddConnection(const std::string& address, const ConnectionType conn_type)
+{
+    assert(conn_type == ConnectionType::OUTBOUND_FULL_RELAY || conn_type == ConnectionType::BLOCK_RELAY);
+
+    int outbound_full_relay = 0;
+    int outbound_block_relay = 0;
+
+    {
+        // count existing connections for each type
+        LOCK(cs_vNodes);
+        for (const CNode* pnode : vNodes) {
+            if (pnode->m_conn_type == ConnectionType::OUTBOUND_FULL_RELAY) {
+                ++outbound_full_relay;
+            } else if (pnode->m_conn_type == ConnectionType::BLOCK_RELAY) {
+                ++outbound_block_relay;
+            }
+        }
+    }
+
+    // max connections of specified type already exist
+    if (conn_type == ConnectionType::OUTBOUND_FULL_RELAY && outbound_full_relay >= m_max_outbound_full_relay) return false;
+    if (conn_type == ConnectionType::BLOCK_RELAY && outbound_block_relay >= m_max_outbound_block_relay) return false;
```

Comment:
You don't need to count the number both connection types if you're only going to test against one. It might also be clearer to use the standard library count_if rather than hand writing the loop.

See what you think of this:

```suggestion
    const int max_connections = conn_type == ConnectionType::OUTBOUND_FULL_RELAY ? m_max_outbound_full_relay :
                                                                                   m_max_outbound_block_relay;

    // Count existing connections
    int existing_connections{0};
    {
        LOCK(cs_vNodes);
        existing_connections = std::count_if(vNodes.begin(),
                                             vNodes.end(),
                                             [conn_type](CNode* node){return node->m_conn_type == conn_type;});
    }

    // Max connections of specified type already exist
    if (existing_connections >= max_connections) return false;

```

Reply:
this is great! thank you! I've incorporated the suggestion